pkg/hubble/relay/observer.go

```diff
@@ -137,28 +121,21 @@ func (s *Server) ServerStatus(ctx context.Context, req *observerpb.ServerStatusR
 	statuses := make(chan *observerpb.ServerStatusResponse, len(peers))
 	for _, p := range peers {
 		p := p
+		if p.conn == nil || p.connErr != nil {
+			s.log.WithField("address", p.Address.String()).Infof(
+				"No connection to peer %s, skipping", p.Name,
+			)
+			go s.connectPeer(p.Name, p.Address.String())
```

Comment:
I think we could wrap this go-routine creation with an if-statement, making sure there is always only one go-routine per peer alive (via atomic boolean. Or in this case, atomic u32 since Go doesn't have atomic booleans):

```go
if atomic.CompareAndSwapUint32(p.isReconnecting, 0, 1) {
  go s.connectPeer(p.Name, p.Address.String())
}
```

`p.isReconnecting` would be of type `*uint32` (i.e. all copies of `hubblePeer` point towards the same boolean). Inside `connectPeer()`, you would have a `defer atomic.StoreUint32(p.isReconnecting, 0)`.

Reply:
I've thought about this when implementing the PR but decided to leave it aside as an optimization to be tackled later. Ultimately, the way peer connections are managed will have to be revisited anyway and adding logic to ensure that there are no concurrent connection attempt to the same peer will make the code more complicated already.

As I think that this PR is already an improvement over the status quo, I propose to defer this for later.